Move shared URL-building logic into a get_url_parts method; implement Page.get_site

gasman opened the pull request:
The `url`, `relative_url` and `full_url` methods of Page all duplicated the logic for matching a page to a site and constructing a relative page path - meaning that a page model with custom URL routing (e.g. blog posts with date-based URLs) would have to override all three of them. This PR moves the shared logic to a new `get_url_parts` method, which returns a tuple of (site_id, site_root_url, relative_page_path); subclasses now only have to override this one method.

In addition, it implements a `get_site` method as requested in #693 (which we would otherwise have to do as a fourth copy of the same logic...)

Fixes #693

kaedroho commented:
Would be nice to add this new method to the model reference docs too: http://docs.wagtail.io/en/v1.2/reference/pages/model_reference.html

Also, if you consider ``get_url_parts`` to be public API, please add tests/docs for that too. If not, perhaps prefix with an undercore to indicate this?

carlmjohnson commented on wagtail/wagtailcore/models.py:
```diff
@@ -625,12 +625,29 @@ def is_navigable(self):
-                return root_url + reverse('wagtail_serve', args=(self.url_path[len(root_path):],))
+        try:
+            site_id, root_url, page_path = self.get_url_parts()
+        except TypeError:
```
My fear is that this will cover up something going wrong in the call. Why not,

```python
url_parts = self.get_url_parts()

# get_url_parts returned None; page is not routable
if not url_parts:
    return

site_id, root_url, page_path = url_parts
return root_url + page_path
```
?

gasman replied:
Good point! I was trying to follow the Python 'EAFP' principle, but that often has the risk of catching more exceptions than you intended, as it does here.

gasman commented:
Rebased and updated (docs / tests added for `get_url_parts`, non-routable page checks changed to avoid catching unrelated `TypeError`s) - ready for review again.

kaedroho commented:
Merged in 7861c2b4929b08b46978e8a4e187fcc3ca92c972 +parents
